tdm/docs/architecture.md

```diff
@@ -0,0 +1,222 @@
+                                                            │
+                                                            │
+                                                       TEE-IO ABI
+                                                            │
+                                                            │
+                                                            ▽
+┏━━━━━━━━━━━━━━━━━━━━━━━┓                    ┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
+┃                       ┃                    ┃                            ┃
+┃  Host VMM/Hypervisor  ┃                    ┃  Trusted Security Manager  ┃
+┃    (KVM+QEMU/CLH)     ┃◁──────TEE ABI─────▷┃   (TDX Module, ASP, etc)   ┃
+┃                       ┃                    ┃                            ┃
+┗━━━━━━━━━━━━━━━━━━━━━━━┛                    ┗━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛
+            △
+            ┃
+        PCIe Bus
+            ┃
+            ┃
+            ▽
+┏━━━━━━━━━━━━━━━━━━━━━━━┓
+┃ TEE-IO Compliant PCIe ┃
+┃        Device         ┃
+┗━━━━━━━━━━━━━━━━━━━━━━━┛
+
+```
+
+In order for the TDM to achieve the above described goals, it must be able to:
+
+1. Detect VMM-assigned PCIe devices. This relies on the Linux kernel `udev`
+   subsystem. TDM also assumes the TSM must not enable any DMA or MMIO mappings
+   before it accepted the device.
+2. Authenticate the assigned device. Both authentication and attestation of the
+   TDI relies on the ability for the TDM to fetching a certificate chain and an
+   attestation evidence it. This is provided by the Linux kernel confidential
+   computing ABI, which is generally exposed as a vendor-specific, ioctl-based
+   character device.
+3. Verify the assigned device. Before accepting a TDI into the confidential
+   guest TCB, the TDM must use the Linux kernel CoCo ABI to verify the following
+   security attributes from the TSM:
+   * Device link state
+   * TDI state
+   * TDI MMIO mappings
+4. Attest of the assigned device trustworthiness. As part of the overall device
+   verification process, the TDM must run a local or remote attestation of the
+   device attestation evidence. To support local attestation, the TDM
+   practically becomes a Relying Party.
+5. Notify the TSM of it TDI acceptance decision. Here as well, this relies on
+   the Linux kernel CoCo ABI.
+
+This leads to a natural split of the TDM software architecture into 4 main
+modules:
+
+1. The **TDM Core** runs the main TDM flows and managing confidential guest
+   assignment with the support of the detection, ABI plugins and attestation
+   modules.
+2. The **Device Detection** module traps device discovery, assignment and
+   removal from the confidential guest PCIe emulated bus.
+3. The **Linux CoCo ABI plugins** abstracts the vendor specific CoCo Linux
+   kernel ABIs in order expose a vendor-agnostic internal interface for the TDM
+   core to consume.
+4. The **Device Attestation** is actually a relying party implementation, for
```

Comment:
If we use local attestation within the guest, how can we set the attestation policy and reference values? Does this mean they need to be built into the rootfs image?

Reply:
When running local attestation, I think that the guest can be provisioned with attestation policies and reference values in two ways:

1. Statically: The guest image is built with these pieces of data.
2. Dynamically: The guest could use an attestation token to retrieve the policy and RV from a relying party and run a local attestation service.

The TDM should support both attestation methods (local and remote) through a TDM configuration file.